jgrapht-core/src/main/java/org/jgrapht/alg/CliqueMinimalSeparatorDecomposition.java

```diff
@@ -0,0 +1,452 @@
+
+            // mark y reached and add y to reach
+            for (V y : Y) {
+                reached.add(y);
+                addToReach(vertexLabels.get(y), y, reach);
+            }
+
+            for (int j = 0; j < graph.vertexSet().size(); j++) {
+                if (!reach.containsKey(j)) {
+                    continue;
+                }
+                while (reach.get(j).size() > 0) {
+                    // remove a vertex y from reach(j)
+                    V y = reach.get(j).iterator().next();
+                    reach.get(j).remove(y);
+
+                    for (V z : Graphs.neighborListOf(gprime, y)) {
+                        if (!reached.contains(z)) {
+                            reached.add(z);
+                            if (vertexLabels.get(z) > j) {
+                                Y.add(z);
+                                E fillEdge = graph.getEdgeFactory().createEdge(
+                                    v,
+                                    z);
+                                fillEdges.add(fillEdge);
+                                addToReach(vertexLabels.get(z), z, reach);
+                            } else {
+                                addToReach(j, z, reach);
+                            }
+                        }
+                    }
+
+                }
+            }
+
+            for (V y : Y) {
+                chordalGraph.addEdge(v, y);
+                vertexLabels.put(y, vertexLabels.get(y) + 1);
+            }
+
+            meo.addLast(v);
+            gprime.removeVertex(v);
+            vertexLabels.remove(v);
+        }
+    }
+
+    /**
+     * Get the vertex with the maximal label.
+     * 
+     * @param vertexLabels Map that gives a label for each vertex.
+     * @return Vertex with the maximal label.
+     */
+    private V getMaxLabelVertex(Map<V, Integer> vertexLabels)
+    {
+        Iterator<Entry<V, Integer>> iterator = vertexLabels.entrySet()
+            .iterator();
+        Entry<V, Integer> max = iterator.next();
+        while (iterator.hasNext()) {
+            Entry<V, Integer> e = iterator.next();
+            if (e.getValue() > max.getValue())
+                max = e;
+        }
+        return max.getKey();
+    }
+
+    /**
+     * Add a vertex to reach.
+     * 
+     * @param k vertex' label
+     * @param v the vertex
+     * @param r the reach structure.
+     */
+    private void addToReach(Integer k, V v, HashMap<Integer, HashSet<V>> r)
+    {
+        if (r.containsKey(k)) {
+            r.get(k).add(v);
+        } else {
+            HashSet<V> set = new HashSet<V>();
+            set.add(v);
+            r.put(k, set);
+        }
+    }
+
+    /**
+     * Compute the unique decomposition of the input graph G (atoms of G).
+     * Implementation of algorithm Atoms as described in Berry et al. (2010),
+     * DOI:10.3390/a3020197, <a href="http://www.mdpi.com/1999-4893/3/2/197">
+     * http://www.mdpi.com/1999-4893/3/2/197</a>
+     */

```

Comment:
This should probably be private since the public methods which need it call it lazily.